src/local.rs

```diff
@@ -1,45 +1,185 @@
+
+                __KEY.get(__init)
+            }
+
+            unsafe {
+                $crate::local::LocalKey::new(__getit)
+            }
+        }
+    };
+
+    ($(#[$attr:meta])* $vis:vis static $name:ident, $t:ty, $init:expr) => {
+        $(#[$attr])* $vis const $name: $crate::local::LocalKey<$t> =
+            $crate::__locality_inner!(@key $(#[$attr])* $vis $name, $t, $init);
+    };
+}
+
+#[cfg(not(feature = "std"))]
+mod no_std {
+    use crate::stdlib::cell::UnsafeCell;
+    use crate::stdlib::{
+        fmt, mem, ptr,
+        sync::atomic::{AtomicPtr, AtomicU8, AtomicUsize, Ordering},
+    };
+    use crate::{Id, Locality};
+
+    pub(crate) static MAX_LOCALITIES: AtomicUsize = AtomicUsize::new(0);
+
+    static DEFAULT_LOCALITY: AtomicPtr<()> = AtomicPtr::new(ptr::null_mut());
+
+    pub struct SetDefaultError {
+        msg: &'static str,
+    }
+
+    impl fmt::Debug for SetDefaultError {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "SetDefaultError: {}", self.msg)
+        }
+    }
+
+    pub fn set_locality<L: Locality>() -> Result<(), SetDefaultError> {
+        let max = L::MAX_LOCALITIES;
+        MAX_LOCALITIES
+            .compare_exchange(0, max, Ordering::AcqRel, Ordering::Acquire)
+            .map_err(|_| SetDefaultError {
+                msg: "default locality was already set!",
+            })?;
+
+        let locality = L::current as *mut ();
+
+        DEFAULT_LOCALITY
+            .compare_exchange(
+                ptr::null_mut(),
+                locality,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            )
+            .map_err(|_| SetDefaultError {
+                msg: "default locality was already set!",
+            })?;
+        Ok(())
+    }
+
+    pub fn reset() {
+        MAX_LOCALITIES.store(0, Ordering::Acquire);
+        DEFAULT_LOCALITY.store(ptr::null_mut(), Ordering::Acquire);
+    }
+
+    #[doc(hidden)]
+    pub struct __Store<T: Sized> {
+        items: UnsafeCell<crate::stdlib::vec::Vec<Option<T>>>,
+        initialized: AtomicU8,
+    }
+
+    unsafe impl<T: Sized> Sync for __Store<T> {}
+
+    impl<T: Sized> __Store<T> {
+        pub const fn new() -> Self {
+            let items = UnsafeCell::new(crate::stdlib::vec::Vec::new());
+            let initialized = AtomicU8::new(0);
+
+            Self { items, initialized }
+        }
+
+        fn initialize(&self) {
+            let size = MAX_LOCALITIES.load(Ordering::Acquire);
+            assert!(size > 0, "you should have called set_locality");
+
+            match self.initialized.compare_and_swap(0, 1, Ordering::Acquire) {
+                2 => return,
+                0 => {
+                    unsafe { &mut *(self.items.get()) }.resize_with(size, || None);
+                    self.initialized.store(2, Ordering::Release);
+                }
+                1 => while self.initialized.load(Ordering::Acquire) != 2 {},
```

Comment:
a dummy spinlock for initialization, should only happen once.

Reply:
that's probably fine